src/backend.c

```diff
@@ -4926,52 +4890,23 @@ int run_cuda_kernel_utf8toutf16le (hashcat_ctx_t *hashcat_ctx, hc_device_param_t
 
   if (hc_cuLaunchKernel (hashcat_ctx, function, num_elements, 1, 1, kernel_threads, 1, 1, 0, device_param->cuda_stream, device_param->kernel_params_utf8toutf16le, NULL) == -1) return -1;
 
-  if (hc_cuStreamSynchronize (hashcat_ctx, device_param->cuda_stream) == -1) return -1;
-
   return 0;
 }
 
-int run_cuda_kernel_memset (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *device_param, CUdeviceptr buf, const u32 value, const u64 size)
+int run_cuda_kernel_memset (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *device_param, CUdeviceptr buf, const u64 offset, const u8 value, const u64 size)
 {
-  const u64 num16d = size / 16;
-  const u64 num16m = size % 16;
-  u32 tmp[4];
-
-  if (num16d)
-  {
-    device_param->kernel_params_memset[0]       = (void *) &buf;
-    device_param->kernel_params_memset_buf32[1] = value;
-    device_param->kernel_params_memset_buf64[2] = num16d;
-
-    const u64 kernel_threads = device_param->kernel_wgs_memset;
-
-    u64 num_elements = CEILDIV (num16d, kernel_threads);
-
-    CUfunction function = device_param->cuda_function_memset;
-
-    if (hc_cuLaunchKernel (hashcat_ctx, function, num_elements, 1, 1, kernel_threads, 1, 1, 0, device_param->cuda_stream, device_param->kernel_params_memset, NULL) == -1) return -1;
-  }
-
-  if (num16m)
-  {
-    tmp[0] = value;
-    tmp[1] = value;
-    tmp[2] = value;
-    tmp[3] = value;
-
-    if (hc_cuMemcpyHtoDAsync (hashcat_ctx, buf + (num16d * 16), tmp, num16m, device_param->cuda_stream) == -1) return -1;
-  }
-
-  if (hc_cuStreamSynchronize (hashcat_ctx, device_param->cuda_stream) == -1) return -1;
+  return hc_cuMemsetD8Async (hashcat_ctx, buf + offset, value, size, device_param->cuda_stream);
+}
 
-  return 0;
+int run_cuda_kernel_memset32 (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *device_param, CUdeviceptr buf, const u64 offset, const u32 value, const u64 size)
+{
+  return hc_cuMemsetD32Async (hashcat_ctx, buf + offset * sizeof (u32), value, size, device_param->cuda_stream);
```

Comment:
The * sizeof (u32) feels unintuitive, because an offset is supposed to be absolute (in bytes). Same applies to the sibling functions for HIP and OpenCL. What does the SDK documentation say about?

Reply:
> The * sizeof (u32) feels unintuitive, because an offset is supposed to be absolute (in bytes). Same applies to the sibling functions for HIP and OpenCL. What does the SDK documentation say about?

Right. Offset param should offset_elem, as SDK says that: 
> The dstDevice pointer must be four byte aligned.

And as buf is allocated by CUDA, I think it's guaranteed to have 256 byte alignment so with current implementation it's impossible not to have four byte aligned.. I see run_cuda_kernel_memset32 as helper routine, as we are not hiding the actual library call hc_cuMemsetD32Async. One option would be to make a single run_cuda_kernel_memset, and check if we can call hc_cuMemsetD32Async or if we need to call hc_cuMemsetD8Async depending on offset and size and in this case those would be bytes, not elements.

Reply:
Also fromclEnqueueFillBuffer() API. It has

> offset
The location in bytes of the region being filled in buffer and must be a multiple of pattern_size.

Reply:
Now that I am thinking this again, you are right and I should have implemented this as a smart helper that automatically picks hc_cuMemsetD32Async or hc_cuMemsetD8Async. I will "fix" this tomorrow.

Reply:
Now run_cuda_kernel_memset32, run_hip_kernel_memset32, run_opencl_kernel_memset32 offset and size are assumed to be bytes and both must be a multiple of element size.  62a06f7